x/dymns/keeper/hooks.go

```diff
@@ -0,0 +1,365 @@
+		// and process next order, to prevent chain reaction when an individual order failed to process
+		errApplyStateChange := osmoutils.ApplyFuncIfNoError(ctx, func(ctx sdk.Context) error {
+			if so.HighestBid == nil {
+				e.DeleteSellOrder(ctx, so.AssetId, dymnstypes.TypeAlias)
+				return nil
+			}
+
+			/**
+			For the Sell-Orders which the assets are prohibited to trade,
+			the Sell-Order will be force cancelled and the bids will be refunded.
+
+			Why some aliases are prohibited to trade? And what are they?
+			In module params, there is a list of alias mapping for some external well-known chains.
+			So those aliases are considered as reserved for the external chains,
+			therefor trading is not allowed.
+
+			Why can someone own a prohibited alias?
+			An alias can be bought before the reservation was made.
+			But when the alias becomes reserved for the external well-known chains,
+			the alias will be prohibited to trade.
+
+			Why can someone place a Sell-Order for the prohibited alias?
+			When a Sell-Order created before the reservation was made.
+			*/
+			if _, forceCancel := prohibitedToTradeAliases[so.AssetId]; forceCancel {
+				// Sell-Order will be force cancelled and refund bids if any,
+				// when the alias is prohibited to trade
+				if err := e.RefundBid(ctx, *so.HighestBid, dymnstypes.TypeAlias); err != nil {
+					logger.Error("failed to refund bid for a force-to-cancel sell order", "asset-id", so.AssetId, "error", err)
+					return err
+				}
+				e.DeleteSellOrder(ctx, so.AssetId, dymnstypes.TypeAlias)
+				return nil
+			}
+
+			if err := e.CompleteAliasSellOrder(ctx, so.AssetId); err != nil {
+				logger.Error("failed to complete sell order", "asset-id", so.AssetId, "error", err)
+				return err
+			}
+
+			return nil
+		})
+
+		if errApplyStateChange == nil {
+			aSoe.Remove(so.AssetId)
+		}
+	}
+
+	if err := e.SetActiveSellOrdersExpiration(ctx, aSoe, dymnstypes.TypeAlias); err != nil {
+		e.Keeper.Logger(ctx).Error("failed to update active SO expiry", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// getFinishedSellOrders returns the finished Sell-Orders for the asset type.
+// Finished Sell-Orders are the Sell-Orders that have expired or completed by having a bid.
+func (e epochHooks) getFinishedSellOrders(ctx sdk.Context, assetType dymnstypes.AssetType, logger log.Logger) (
+	finishedSellOrders []dymnstypes.SellOrder,
+	aSoe *dymnstypes.ActiveSellOrdersExpiration,
+) {
+	aSoe = e.Keeper.GetActiveSellOrdersExpiration(ctx, assetType)
+	if len(aSoe.Records) < 1 {
+		return
+	}
+
+	blockEpochUTC := ctx.BlockTime().Unix()
+	logger = logger.With("asset-type", assetType.FriendlyString(), "time", blockEpochUTC)
+
+	invalidRecordsToRemove := make([]string, 0)
+
+	for i, record := range aSoe.Records {
+		if record.ExpireAt > blockEpochUTC {
+			// skip not expired ones
+			continue
+		}
+
+		so := e.GetSellOrder(ctx, record.AssetId, assetType)
```

Comment:
It is used to detect the Invalid State records. As described in the comment:

The following part is to filter out the finished Sell-Orders,
plus one logic to correct the invalid Sell-Order expiry.
Currently, there is no such case that can cause an invalid Sell-Order expiry,
but it's good to have a correction logic for the future
because this part is executed by hook, and to make sure the hook always be executed efficiently,
we need to make sure the data is always valid.